### PR TITLE
network: Add `IgnoreDF=` option for Fragmentation control

### PR DESCRIPTION
**Pull request by vfreex**
From `ip-link(8)`:
> [no]ignore-df - enables/disables IPv4 DF
                      suppression on this tunnel.  Normally datagrams
                      that exceed the MTU will be fragmented; the
                      presence of the DF flag inhibits this, resulting
                      instead in an ICMP Unreachable (Fragmentation
                      Required) message.  Enabling this attribute causes
                      the DF flag to be ignored.

If this option is enabled for a GRE/GRETAP tunnel, the `DF` flag in the outer IP header will not inherit the inner IP header's `DF` flag.

This is useful to transfer packets that exceed the MTU of the underlay network.

### CODE REVIEW

File: man/systemd.netdev.xml [hunk 1]
```diff
@@ -1223,6 +1223,15 @@
           the tunnel.</para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>IgnoreDF=</varname></term>
+        <listitem>
+          <para>Takes a boolean. When true, enables IPv4 DF flag suppression on
+          the tunnel. Note that if <varname>IgnoreDF=</varname> is true,
+          <varname>DiscoverPathMTU=</varname> must be set to false.
+          Only applicable to GRE and GRETAP tunnels. Defaults to false.</para>
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Is the default value for DiscoverPathMTU false? The docs should really say that.

Maybe reword this as "Note that if IgnoreDF= is set to true, DiscoverPathMTU= cannot be set to false." (This way we don't imply that DiscoverPathMTU=false must be used.)

Please also expand DF. (See how DSCP is described below.)

File: src/network/netdev/tunnel.c [hunk 2]
```diff
@@ -745,7 +749,9 @@ static int netdev_tunnel_verify(NetDev *netdev, const char *filename) {
                 return log_netdev_error_errno(netdev, SYNTHETIC_ERRNO(EINVAL),
                                               "The local address cannot be '%s' when Independent= or AssignToLoopback= is enabled, ignoring.",
                                               strna(netdev_local_address_type_to_string(t->local_type)));
-
+        if (t->ignore_df && t->pmtudisc)
+                return log_netdev_error_errno(netdev, SYNTHETIC_ERRNO(EINVAL),
+                                              "IgnoreDF cannot be enabled when DiscoverPathMTU= is enabled, ignoring.");
```
Comment (keszybz) [inline src/network/netdev/tunnel.c]: Strike ", ignoring.".

All the other messages in this function are wrong too: when we return an error, we shouldn't say "ignoring".
It'd be nice if you fixed that (in a separate commit message).

**Review by keszybz — CHANGES_REQUESTED** [review]
Looks reasonable.

File: src/network/netdev/tunnel.c [hunk 3]
```diff
@@ -745,7 +749,9 @@ static int netdev_tunnel_verify(NetDev *netdev, const char *filename) {
                 return log_netdev_error_errno(netdev, SYNTHETIC_ERRNO(EINVAL),
                                               "The local address cannot be '%s' when Independent= or AssignToLoopback= is enabled, ignoring.",
                                               strna(netdev_local_address_type_to_string(t->local_type)));
-
+        if (t->ignore_df && t->pmtudisc)
```
Comment (yuwata) [inline src/network/netdev/tunnel.c]: How about to make `pmtudisc` tristate, then refuse to configure the tunnel when `DiscoverPathMTU=` is explicitly enabled?
```c
if (t->pmtudisc > 0 && t->ignore_df)
        return log_netdev_error_errno(...);
if (t->pmtudisc < 0)
        t->pmtudisc = !t->ignore_df;
```

File: src/network/netdev/tunnel.c [hunk 4]
```diff
@@ -1190,6 +1196,7 @@ static void netdev_tunnel_init(NetDev *netdev) {
 
         t->local_type = _NETDEV_LOCAL_ADDRESS_TYPE_INVALID;
         t->pmtudisc = true;
+        t->ignore_df = false;
```
Comment (yuwata) [inline src/network/netdev/tunnel.c]: Please drop unnecessary initialization. The object is already initialized with zero.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: man/systemd.netdev.xml (hunk 1 above)
Comment (vfreex) [reply man/systemd.netdev.xml]: The default value of `DiscoverPathMTU` is true (https://github.com/systemd/systemd/blob/f2b98d59295c9e9d81a4e7f29122bb6e70e6b37b/src/network/netdev/tunnel.c#L1198). To set `IgnoreDF=` to true, `DiscoverPathMTU=` must be set to false.

I will expanded DF in the next push

File: src/network/netdev/tunnel.c (hunk 4 above)
Comment (vfreex) [reply src/network/netdev/tunnel.c]: done

File: src/network/netdev/tunnel.c (hunk 3 above)
Comment (vfreex) [reply src/network/netdev/tunnel.c]: done. please have another look.

File: src/network/netdev/tunnel.c (hunk 2 above)
Comment (vfreex) [reply src/network/netdev/tunnel.c]: done

**Comment by vfreex** [comment]
Thanks for the review! I've updated the PR. Please have another look.

File: man/systemd.netdev.xml [hunk 5]
```diff
@@ -1220,7 +1220,19 @@
...
         <listitem>
           <para>Takes a boolean. When true, enables Path MTU Discovery on
-          the tunnel.</para>
+          the tunnel.
+          When <varname>IgnoreDF=</varname> is explicitly set to true,
+          defaults to false. Otherwise, defaults to true.</para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>IgnoreDF=</varname></term>
```
Comment (poettering) [inline man/systemd.netdev.xml]: maybe spell this out as IgnoreDontFragment=? no need to be needlessly cryptic. makes the .netdev file a lot more self-explanatory, people wouldn't have to read up the docs anymore to see wtf this is.

Comment (vfreex) [reply man/systemd.netdev.xml]: Makes sense. Renamed.

File: man/systemd.netdev.xml [hunk 6]
```diff
@@ -1220,7 +1220,19 @@
         <term><varname>DiscoverPathMTU=</varname></term>
         <listitem>
           <para>Takes a boolean. When true, enables Path MTU Discovery on
-          the tunnel.</para>
+          the tunnel.
+          When <varname>IgnoreDontFragment=</varname> is explicitly set to true,
```
Comment (yuwata) [inline man/systemd.netdev.xml]: `is explicitly set to true` -> `is enabled` is simpler??

File: src/network/netdev/tunnel.c [hunk 7]
```diff
@@ -746,6 +750,11 @@ static int netdev_tunnel_verify(NetDev *netdev, const char *filename) {
                                               "The local address cannot be '%s' when Independent= or AssignToLoopback= is enabled, ignoring.",
                                               strna(netdev_local_address_type_to_string(t->local_type)));
 
+        if (t->pmtudisc > 0 && t->ignore_df)
+                return log_netdev_error_errno(netdev, SYNTHETIC_ERRNO(EINVAL),
+                                              "IgnoreDontFragment cannot be enabled when DiscoverPathMTU= is enabled");
```
Comment (yuwata) [inline src/network/netdev/tunnel.c]: `IgnoreDontFragment` -> `IgnoreDontFragment=` (i.e. missing `=`)

File: man/systemd.netdev.xml [hunk 8]
```diff
@@ -1220,7 +1220,19 @@
...
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>IgnoreDontFragment=</varname></term>
+        <listitem>
+          <para>Takes a boolean. When true, enables IPv4 Don't Fragment (DF) suppression on
+          the tunnel. Defaults to false.
+          Note that if <varname>IgnoreDontFragment=</varname> is set to true,
+          <varname>DiscoverPathMTU=</varname> cannot be set to true.
+          Only applicable to GRE and GRETAP tunnels.</para>
```
Comment (yuwata) [inline man/systemd.netdev.xml]: Is this ignored for erspan?

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks mostly good.
Please add a simple test in `test/test-network/systemd-networkd-tests.py`.

File: man/systemd.netdev.xml (hunk 8 above)
Comment (vfreex) [reply man/systemd.netdev.xml]: yes, it works for erspan as well. I've updated the description and included a simple test case.

File: src/network/netdev/tunnel.c (hunk 7 above)
Comment (vfreex) [reply src/network/netdev/tunnel.c]: done

File: man/systemd.netdev.xml (hunk 6 above)
Comment (vfreex) [reply man/systemd.netdev.xml]: done

**Review by yuwata — APPROVED** [review]
LGTM.
